plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/RestReindexAction.java

```diff
@@ -0,0 +1,167 @@
+ * under the License.
+ */
+
+package org.elasticsearch.plugin.reindex;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.WriteConsistencyLevel;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.ObjectParser.ValueType;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.VersionType;
+import org.elasticsearch.index.query.QueryParseContext;
+import org.elasticsearch.indices.query.IndicesQueriesRegistry;
+import org.elasticsearch.rest.BaseRestHandler;
+import org.elasticsearch.rest.BytesRestResponse;
+import org.elasticsearch.rest.RestChannel;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.script.Script;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.plugin.reindex.ReindexAction.INSTANCE;
+import static org.elasticsearch.rest.RestRequest.Method.POST;
+import static org.elasticsearch.rest.RestStatus.BAD_REQUEST;
+
+/**
+ * Expose IndexBySearchRequest over rest.
+ */
+public class RestReindexAction extends BaseRestHandler {
+    private static final ObjectParser<ReindexRequest, QueryParseContext> PARSER = new ObjectParser<>("reindex");
+    static {
+        ObjectParser.Parser<SearchRequest, QueryParseContext> sourceParser = (parser, search, context) -> {
+            try {
+                /*
+                 * Extract the parameters that we need from the parser. We could
+                 * do away with this hack when search source has an
+                 * ObjectParser.
+                 */
+                Map<String, Object> source = parser.map();
+                Object index = source.remove("index");
+                if (index != null) {
+                    if (index instanceof List) {
+                        @SuppressWarnings("unchecked")
+                        List<String> list = (List<String>) index;
+                        search.indices(list.toArray(new String[list.size()]));
+                    } else if (index instanceof String) {
+                        search.indices((String) index);
+                    } else {
+                        throw new IllegalArgumentException("Expected index to be a list of a string but was [" + index + ']');
+                    }
+                }
+                Object type = source.remove("type");
+                if (type != null) {
```

Comment:
this code looks structurally similar to the index handling code. Does it make sense to create a separate method for that?